cdap-security/src/test/java/co/cask/cdap/security/server/ExternalAuthenticationServerMutualSSLTest.java

```diff
@@ -0,0 +1,264 @@
+import java.net.InetAddress;
+import java.net.URL;
+import java.security.KeyStore;
+import java.security.SecureRandom;
+import java.security.cert.CertificateException;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509TrustManager;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests for {@link ExternalAuthenticationServer} with 2-way SSL enabled.
+ */
+public class ExternalAuthenticationServerMutualSSLTest extends ExternalAuthenticationServerTestBase {
+
+    protected static String getAuthServerHost() {
+        return "127.0.0.1";
+    }
+
+    @BeforeClass
+    public static void beforeClass() throws Exception {
+
+        // adding a new server side key store which has a 10 year long validity
+        URL certUrl = ExternalAuthenticationServerSSLTest.class.getClassLoader().getResource("server-key.jks");
+        Assert.assertNotNull(certUrl);
+
+        // create server's trust store
+        URL trustUrl = ExternalAuthenticationServerSSLTest.class.getClassLoader().getResource("server-trust.jks");
+        Assert.assertNotNull(trustUrl);
+
+        String authHandlerConfigBase = Constants.Security.AUTH_HANDLER_CONFIG_BASE;
+
+        CConfiguration cConf = CConfiguration.create();
+        SConfiguration sConf = SConfiguration.create();
+        cConf.set(Constants.Security.AUTH_SERVER_BIND_ADDRESS, getAuthServerHost());
+        cConf.set(Constants.Security.SSL_ENABLED, "true");
+        cConf.set(Constants.Security.AuthenticationServer.SSL_PORT, "0");
+        cConf.set(authHandlerConfigBase.concat("useLdaps"), "true");
+        cConf.set(authHandlerConfigBase.concat("ldapsVerifyCertificate"), "false");
+        sConf.set(Constants.Security.AuthenticationServer.SSL_KEYSTORE_PATH, certUrl.getPath());
+        configuration = cConf;
+        sConfiguration = sConf;
+
+        String keystorePassword = sConf.get(Constants.Security.AuthenticationServer.SSL_KEYSTORE_PASSWORD);
+        KeyStoreKeyManager keyManager = new KeyStoreKeyManager(certUrl.getFile(), keystorePassword.toCharArray());
+
+
+        // setting 2-way SSL
+        cConf.set(Constants.Security.CLIENT_AUTH_ENABLED, "true");
+        cConf.set(Constants.Security.AuthenticationServer.SSL_TRUSTSTORE_PATH, trustUrl.getPath());
+        // using the same password as the keystore
+        cConf.set(Constants.Security.AuthenticationServer.SSL_TRUSTSTORE_PASSWORD,
+                sConf.get(Constants.Security.AuthenticationServer.SSL_KEYSTORE_PASSWORD));
+
+        TrustStoreTrustManager trustManager = new TrustStoreTrustManager(trustUrl.getPath(),
+                sConf.get(Constants.Security.AuthenticationServer.SSL_KEYSTORE_PASSWORD).toCharArray(),
+                Constants.Security.AuthenticationServer.DEFAULT_SSL_KEYSTORE_TYPE, false);
+
+        SSLUtil sslUtil = new SSLUtil(keyManager, trustManager);
+        ldapListenerConfig = InMemoryListenerConfig.createLDAPSConfig("LDAP",
```

Comment:
I think it would be a good idea to use client certificate authenticator in this test case. This would prevent any issues due to regression in future. What do you say?